src/codec.rs

```diff
@@ -0,0 +1,303 @@
+                        .collect::<Result<Vec<EncodedSignature>, Self::Error>>()?,
+                )
+            },
+            aad: None,
+            ciphertext: None,
+            iv: None,
+            protected: None,
+            recipients: None,
+            tag: None,
+            unprotected: None,
+        })
+    }
+}
+
+impl TryFrom<Encoded> for JsonWebSignature {
+    type Error = Error;
+
+    fn try_from(value: Encoded) -> Result<Self, Self::Error> {
+        let link = Cid::try_from(value.payload.as_ref().ok_or(Error::NotJws)?.as_slice())?;
+        Ok(Self {
+            payload: value.payload.to_base64().ok_or(Error::NotJws)?,
+            signatures: value
+                .signatures
+                .unwrap_or_else(|| Vec::new())
+                .drain(..)
+                .map(Signature::from)
+                .collect(),
+            link,
+        })
+    }
+}
+
+impl<'a> TryFrom<&'a JsonWebEncryption> for Encoded {
+    type Error = Error;
+
+    fn try_from(value: &'a JsonWebEncryption) -> Result<Self, Self::Error> {
+        Ok(Self {
+            payload: None,
+            signatures: None,
+            iv: if value.iv.is_empty() {
+                None
+            } else {
+                Some(Bytes::from_base64(value.iv.as_str())?)
+            },
+            aad: Option::from_base64(value.aad.as_ref().map(|v| v.as_str()))?,
+            tag: if value.tag.is_empty() {
+                None
+            } else {
+                Some(Bytes::from_base64(value.tag.as_str())?)
+            },
+            protected: if value.protected.is_empty() {
+                None
+            } else {
+                Some(Bytes::from_base64(value.protected.as_str())?)
+            },
+            ciphertext: Some(Bytes::from_base64(value.ciphertext.as_str())?),
+            recipients: if value.recipients.is_empty() {
+                None
+            } else {
+                Some(
+                    value
+                        .recipients
+                        .iter()
+                        .map(EncodedRecipient::try_from)
+                        .collect::<Result<Vec<EncodedRecipient>, Self::Error>>()?,
+                )
+            },
+            unprotected: if value.unprotected.is_empty() {
+                None
+            } else {
+                Some(value.unprotected.to_owned())
+            },
+        })
+    }
+}
+
+impl TryFrom<Encoded> for JsonWebEncryption {
+    type Error = Error;
+
+    fn try_from(value: Encoded) -> Result<Self, Self::Error> {
+        Ok(Self {
+            aad: value.aad.to_base64(),
+            ciphertext: value.ciphertext.to_base64().ok_or(Error::NotJwe)?,
+            iv: value.iv.to_base64().ok_or(Error::NotJwe)?,
+            protected: value.protected.to_base64().ok_or(Error::NotJwe)?,
+            recipients: value
+                .recipients
+                .unwrap_or_else(|| Vec::new())
+                .drain(..)
```

Comment:
We do not reuse recipients so draining it means we take ownership of its values as we iterate over them. Otherwise we would have to iterate over references to its vales and clone the values before converting. By draining we are more efficient and explicit we are consuming the vector.